api/tech/oras/client.go

```diff
@@ -0,0 +1,173 @@
+package oras
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	oraserr "oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+type ClientOptions struct {
+	Client    *auth.Client
+	PlainHTTP bool
+}
+
+type Client struct {
+	Client    *auth.Client
+	PlainHTTP bool
+	Ref       string
+}
+
+var (
+	_ Resolver = &Client{}
+	_ Fetcher  = &Client{}
+	_ Pusher   = &Client{}
+	_ Lister   = &Client{}
+)
+
+func New(opts ClientOptions) *Client {
+	return &Client{Client: opts.Client, PlainHTTP: opts.PlainHTTP}
+}
+
+func (c *Client) Resolve(ctx context.Context, ref string) (string, ociv1.Descriptor, error) {
+	src, err := c.createRepository(ref)
+	if err != nil {
+		return "", ociv1.Descriptor{}, err
+	}
+
+	// We try to first resolve a manifest.
+	// _Note_: If there is an error like not found, but we know that the digest exists
+	// we can add src.Blobs().Resolve in here. If we do that, note that
+	// for Blobs().Resolve `not found` is actually `invalid checksum digest format`.
+	// Meaning it will throw that error instead of not found.
+	desc, err := src.Resolve(ctx, ref)
+	if err != nil {
+		if errors.Is(err, oraserr.ErrNotFound) {
+			return "", ociv1.Descriptor{}, errdefs.ErrNotFound
+		}
+
+		return "", ociv1.Descriptor{}, fmt.Errorf("failed to resolve manifest %q: %w", ref, err)
+	}
+
+	return "", desc, nil
+}
+
+func (c *Client) Fetcher(ctx context.Context, ref string) (Fetcher, error) {
+	c.Ref = ref
+	return c, nil
+}
+
+func (c *Client) Pusher(ctx context.Context, ref string) (Pusher, error) {
+	c.Ref = ref
+	return c, nil
+}
+
+func (c *Client) Lister(ctx context.Context, ref string) (Lister, error) {
+	c.Ref = ref
+	return c, nil
+}
+
+func (c *Client) Push(ctx context.Context, d ociv1.Descriptor, src Source) error {
+	reader, err := src.Reader()
+	if err != nil {
+		return err
+	}
+
+	repository, err := c.createRepository(c.Ref)
+	if err != nil {
+		return err
+	}
+
+	if split := strings.Split(c.Ref, ":"); len(split) == 2 {
```

Comment:
shouldnt the reference have an IsTag() method somewhere? maybe this is too low level already here, so feel free to ignore if it doesnt apply.

Reply:
It should but it doesn't. 🤣🤣